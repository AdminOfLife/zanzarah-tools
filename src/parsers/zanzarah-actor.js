const jBinary = require('jbinary');

const {
  Switch,
  DumpContext,
  DynamicString,
} = require('../utils/jbinary');

const ArraySections = [
  'AnimationFilename_Body',
  'AnimationFilename_Wings',
];

exports.typeSet = {
  'jBinary.all': 'Actor',
  'jBinary.littleEndian': true,

  Switch,
  DumpContext,
  DynamicString,

  Actor: jBinary.Template({
    baseType: ['array', 'Section'],
    read() {
      // Remove first and last sections - header and eos
      const sections = this.baseRead().slice(1, -1);
      // Index sections by name, collapse those with identical names to arrays
      return sections.reduce((result, section) => {
        const key = section.name.slice(1, -1);
        if (ArraySections.includes(key)) {
          result[key] = result[key] || [];
          result[key].push(section.data);
        } else {
          result[key] = section.data;
        }
        return result;
      }, {});
    },
    write(value) {
      const sectionNames = ['ActorExDescriptionFileData'];
      sectionNames.push(...Object.keys(value));
      sectionNames.push('EOS');
      const sections = [];
      sectionNames.forEach(name => {
        const items = ArraySections.inclues(name) ? value[name] : [value[name]];
        sections.push(...items.map(data => ({
          name: `[${name}]`,
          data,
        })));
      });
      return this.baseWrite(sections);
    },
  }),

  Section: ['extend', {
    name: 'DynamicString',
    data: ['Switch', context => {
      const handler = `${context.name.slice(1, -1)}Data`;
      if (!exports.typeSet[handler]) {
        throw new Error(`Unsupported section: ${context.name}`);
      }
      return handler;
    }],
  }, {
    _: ['DumpContext', null, 1],
  }],

  ActorExDescriptionFileData: {},
  HeadBoneIDData: 'int32',
  EffectBoneIDData: 'int32',
  AttachWingsToBoneData: 'int32',
  ModelFilename_BodyData: DynamicString,
  ModelFilename_WingsData: DynamicString,
  AnimationPoolID_BodyData: 'int32',
  AnimationPoolID_WingsData: 'int32',
  AnimationFilename_BodyData: {
    fileName: DynamicString,
    id: 'int32',
  },
  AnimationFilename_WingsData: {
    fileName: DynamicString,
    id: 'int32',
  },
  EOSData: {},
};
