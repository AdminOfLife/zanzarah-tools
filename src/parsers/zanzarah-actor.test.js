import jBinary from 'jbinary';
import { typeSet } from './zanzarah-actor';

const parse = async fileNameOrBuffer => {
  const binary = await jBinary.load(fileNameOrBuffer, typeSet);
  return binary.readAll();
};

describe('parsers - zanzarah actor', () => {
  it('exports typeSet to parse', () => {
    expect(typeSet).toBeDefined();
  });
  it('parses amy definition', async () => {
    expect(await parse('./public/fixtures/actor-amy.aed')).toMatchSnapshot();
  });
  it('parses random fairy definition', async () => {
    expect(await parse('./public/fixtures/actor-fairy.aed')).toMatchSnapshot();
  });
});
