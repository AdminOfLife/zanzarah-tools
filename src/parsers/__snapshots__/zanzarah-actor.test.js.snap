// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parsers - zanzarah actor parses amy definition 1`] = `
Object {
  "AnimationFilename_Body": Array [
    Object {
      "fileName": "00ast.ska",
      "id": 34,
    },
    Object {
      "fileName": "00blw.ska",
      "id": 31,
    },
    Object {
      "fileName": "00cat0.ska",
      "id": 24,
    },
    Object {
      "fileName": "00cat1.ska",
      "id": 23,
    },
    Object {
      "fileName": "00cat2.ska",
      "id": 19,
    },
    Object {
      "fileName": "00fal0.ska",
      "id": 7,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 5,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 4,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 9,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 10,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 6,
    },
    Object {
      "fileName": "00fwd.ska",
      "id": 2,
    },
    Object {
      "fileName": "00idl.ska",
      "id": 0,
    },
    Object {
      "fileName": "00idl1.ska",
      "id": 11,
    },
    Object {
      "fileName": "00idl2.ska",
      "id": 12,
    },
    Object {
      "fileName": "00jmp.ska",
      "id": 1,
    },
    Object {
      "fileName": "00joy.ska",
      "id": 29,
    },
    Object {
      "fileName": "00she.ska",
      "id": 32,
    },
    Object {
      "fileName": "00sta0.ska",
      "id": 30,
    },
    Object {
      "fileName": "00sta1.ska",
      "id": 38,
    },
    Object {
      "fileName": "00stp.ska",
      "id": 37,
    },
    Object {
      "fileName": "00sup.ska",
      "id": 20,
    },
    Object {
      "fileName": "00sup1.ska",
      "id": 35,
    },
    Object {
      "fileName": "00tal0.ska",
      "id": 13,
    },
    Object {
      "fileName": "00tal1.ska",
      "id": 15,
    },
    Object {
      "fileName": "00tel.ska",
      "id": 6,
    },
    Object {
      "fileName": "00tele.ska",
      "id": 3,
    },
    Object {
      "fileName": "00wal0.ska",
      "id": 17,
    },
    Object {
      "fileName": "00wal1.ska",
      "id": 18,
    },
  ],
  "AnimationPoolID_Body": 0,
  "AnimationPoolID_Wings": -1,
  "AttachWingsToBone": 0,
  "EffectBoneID": 46,
  "HeadBoneID": 7,
  "ModelFilename_Body": "chr01.DFF",
  "ModelFilename_Wings": "",
}
`;

exports[`parsers - zanzarah actor parses random fairy definition 1`] = `
Object {
  "AnimationFilename_Body": Array [
    Object {
      "fileName": "s00bck.ska",
      "id": 24,
    },
    Object {
      "fileName": "s00dance.ska",
      "id": 6,
    },
    Object {
      "fileName": "s00fwd.ska",
      "id": 23,
    },
    Object {
      "fileName": "s00idl00.ska",
      "id": 20,
    },
    Object {
      "fileName": "s00lft.ska",
      "id": 25,
    },
    Object {
      "fileName": "s00rgt.ska",
      "id": 26,
    },
    Object {
      "fileName": "s00sid00.ska",
      "id": 0,
    },
    Object {
      "fileName": "s00wbk.ska",
      "id": 5,
    },
    Object {
      "fileName": "s00wfw.ska",
      "id": 2,
    },
    Object {
      "fileName": "s00wlf.ska",
      "id": 10,
    },
    Object {
      "fileName": "s00wrt.ska",
      "id": 9,
    },
  ],
  "AnimationFilename_Wings": Array [
    Object {
      "fileName": "f000s00w.ska",
      "id": 0,
    },
  ],
  "AnimationPoolID_Body": 3,
  "AnimationPoolID_Wings": 2,
  "AttachWingsToBone": 6,
  "EffectBoneID": 31,
  "HeadBoneID": 7,
  "ModelFilename_Body": "f000s00m.DFF",
  "ModelFilename_Wings": "f000s00w.DFF",
}
`;
